### PR TITLE
Systematic hyperparameter optimisation

### PR DESCRIPTION
**Pull request by TammoR**
*Description of changes:*

New comprehensive HPO functionality. We can now do grid search / random search over hyperparameters. They hps are passed as dictionary containing dictionaries for global parameters and for input columns.

This may need some more refinement before going into master, but this pull request can serve to gather some feedback. Best starting point is to look at the new tests.

-- 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: datawig/simple_imputer.py [hunk 1]
```diff
@@ -130,21 +131,199 @@ def check_data_types(self, data_frame: pd.DataFrame) -> None:
         logger.info("Assuming {} string input columns: {}".format(len(self.string_columns),
                                                                   ", ".join(self.string_columns)))
 
+    def fit_hpo_new(self,
```
Comment (andrey-tpt) [inline datawig/simple_imputer.py]: This method looks very heavy. Consider splitting in a small methods where you add comments to explain what's going on. 

I'd rather see something like this or similar 

```
def method():
  set_defaults()
  validated_inputs()
  for conf in configurations:
    try_conf() 
```

Comment (TammoR) [reply datawig/simple_imputer.py]: Yes, very good point.

**Comment by TammoR** [comment]
Let me be more specific on what this was set out to do and where it falls short. I'll start with illustrating why this is difficult.
We want to be able to optimise all configurations of hyperparameters. This includes parameters with conditional dependencies. E.g. one hyperparameter for a particular column could be the column type: (categorical, string). If the column type is string, a next hyperparameter (that applies only conditionally on the previous one) is the embedding type (Tfidf, Hashing, LSTM). Each of these can again have a different set of conditional hyperparameters, and so on. There would be yet another parameter on top, that is the option to have multiple embeddings for the same column.

It's natural to have a multi-level nested data-structure to store the different parameter values or distributions to draw from. This is done in the proposed implementation, but to keep things simple there are some limitations listed in the following.

- Only one column embedding strategy is supported. This is hard coded (hashing vectoriser at the moment. Maybe swap for tfidf)
- If the same column is embedded multiple times, these embeddings share all hyperparameters except for the token type (chars, words).
- All input columns have a separate column encoder. There is no option to combine different columns into one (other than through pre-processing, which is easy enough)
- There is no mechanism to automatically distinguish text from categorical columns. For categorical encodings, the user needs to supply the column type as parameter.
- The same column can not be encoded as multiple types. E.g. string and categorical.

I'd be interested in whether you think this is a reasonable compromise.

File: datawig/utils.py [hunk 2]
```diff
@@ -45,6 +46,31 @@
 logger.setLevel("INFO")
 
 
+def merge_two_dicts(x, y):
```
Comment (andrey-tpt) [inline datawig/utils.py]: Why is this method required when we already have similar ? https://github.com/awslabs/datawig/blob/master/datawig/utils.py#L60

Comment (TammoR) [reply datawig/utils.py]: It isn't. Will fix in the next commit.

**Pull request closed without merging**